webapp/app/crops/routes.py

```diff
@@ -0,0 +1,390 @@
+        )
+        .join(CropTypeClass, CropTypeClass.id == BatchClass.crop_type_id)
+        .outerjoin(BatchEventClass, BatchEventClass.batch_id == BatchClass.id)
+        .outerjoin(HarvestClass, HarvestClass.batch_event_id == BatchEventClass.id)
+        .outerjoin(LocationClass, LocationClass.id == BatchEventClass.location_id)
+        .filter(
+            and_(
+                BatchEventClass.event_time >= dt_from,
+                BatchEventClass.event_time <= dt_to,
+            )
+        )
+    )
+    df_raw = pd.read_sql(query.statement, query.session.bind)
+
+    # Group events by batch, and manually collect the relevant information from each
+    # event type, if available. Build a new dataframe with only the information we want,
+    # constructing each row as a dictionary first. TODO Might there be a way to lift
+    # some of this work over to the Postgres server? Could be faster.
+    grouped = df_raw.groupby("batch_id")
+    rows = []
+    for batch_id, group in grouped:
+        details = collect_batch_details(group)
+        if details is not None:
+            details["batch_id"] = batch_id
+            rows.append(details)
+
+    df = pd.DataFrame(
+        rows,
+        columns=[
+            "batch_id",
+            "last_event",
+            "crop_type_id",
+            "crop_type_name",
+            "tray_size",
+            "number_of_trays",
+            "weigh_time",
+            "propagate_time",
+            "transfer_time",
+            "harvest_time",
+            "location_id",
+            "zone",
+            "aisle",
+            "column",
+            "shelf",
+            "location",
+            "crop_yield",
+            "waste_disease",
+            "waste_defect",
+            "over_production",
+        ],
+    )
+    # Format the time strings. Easier to do here than in the Jinja template.
+    for column in ["weigh_time", "propagate_time", "transfer_time", "harvest_time"]:
+        df[column] = pd.to_datetime(df[column]).dt.strftime("%Y-%m-%d %H:%M")
+    results_arr = df.to_dict("records")
+
+    if request.method == "POST":
+        return download_csv(results_arr, "batch_list")
+    else:
+        return render_template(
+            "batch_list.html",
+            batches=results_arr,
+            dt_from=dt_from.strftime("%B %d, %Y"),
+            dt_to=dt_to.strftime("%B %d, %Y"),
+        )
+
+
+def distance_metric(location1, location2):
+    """Compute a custom notion distance between two locations.
+
+    The two locations should be tuples of (zone, aisle, column, shelf).
+
+    The return value is a float.
+    """
+    zone1, aisle1, column1, shelf1 = location1
+    zone2, aisle2, column2, shelf2 = location2
+    distance = 0.0
+    # TODO Improve this metric.
+    if zone1 != zone2:
+        distance += 1000.0
+    if aisle1 != aisle2:
+        distance += 1
+    distance += abs(column1 - column2)
+    distance += abs(shelf1 - shelf2)
+    return distance
+
```

Comment:
I guess this is something that we would eventually like to improve by having a proper coordinate system for the farm.   We can maybe tweak sooner by estimating the distance between aisles (though this doesn't matter anyway if everything is in Tunnel 4), and the relative horizontal vs vertical distance between trays (though this is also not obvious, as probably the height is more important than horizontal position in determining temp+humidity).    
So (long story short) I think this is fine for now :)

Reply:
Agree with this. I think height might actually be substantially more important, though I'm unsure if we want to take that into account. If we do, then there's probably no way to avoid some made up constants being involved in the metric (which currently are there, but could be avoided if we used the dimensions of the shelves).